src/vaadin-grid-data-provider-mixin.html

```diff
@@ -326,6 +326,10 @@
 
             this._increasePoolIfNeeded(0);
           }
+
+          if (this._dataProviderDone) {
+            this._dataProviderDone();
```

Comment:
At this point we can't be sure if the grid is actually visible: if might be on another view of invisible for other reasons. Essentially the following test (needs cleanup/dry) should also pass:
```
it('should have correct column widths once visible', done => {
  grid = fixture('auto-width');
  grid.hidden = true;
  grid.items = [
    {a: 'fubar', b: 'foo', c: 'foo', d: 'a'},
    {a: 'foo', b: 'foo bar baz', c: 'foo', d: 'bar'},
    {a: 'foo', b: 'foo baz', c: 'foo bar', d: 'baz'},
  ];
  // Allowed margin of measurement to keep the test from failing if there are small differences in rendered text
  // width on different platforms or if there are small changes to styles which affect horizontal margin/padding.
  const delta = 5;

  requestAnimationFrame(() => {
    grid.hidden = false;

    // recalculateColumnWidths() is run in requestAnimationFrame
    requestAnimationFrame(() => {
      expect(parseInt(columns[0].width)).to.be.closeTo(71, delta);
      expect(parseInt(columns[1].width)).to.be.closeTo(114, delta);
      expect(parseInt(columns[2].width)).to.be.closeTo(84, delta);
      expect(parseInt(columns[3].width)).to.be.closeTo(107, delta);

      done();
    });

  });
});
```

Another thing we can't yet know here is if the grid actually has rendered rows (`_effectiveSize`). There's a check above that uses `!this._cache.isLoading()`, so if there's a lazy data provider that's still waiting for data to the second page when the data for the first page has already arrived, there might not be any physical rows in the DOM. This case also needs a test case.

Running something like the following here in the data provider mixin _and_ after "animationend" (grid is actually visible) should work.
```
if (!this._initialColumnWidthSet && Array.from(this.$.items.children).filter(row => row.clientHeight).length) {
  this._initialColumnWidthSet = true;
  this.requestAnimationFrame(() => this.recalculateColumnWidths());
}
```

Reply:
Thanks for the input. I improved the logic and tests a bit. Please take another look.

Reply:
Instead of `filter()` in the suggested `Array.from(this.$.items.children).filter(row => row.clientHeight).length` I wanted to use `find()` to avoid reading `clientHeight` of all rows unnecessarily but stop when it find the first one with a truthy `clientHeight`. But since IE11 doesn't have `Array.prototype.find` I made a small helper method `__hasRenderedRows()` which basically does the same logic in a `for..of` loop returning when it find the first matching row. This also makes the `if` condition more readable as `if (!this._initialColumnWidthsSet && this.__hasRenderedRows()) {`

Reply:
`for..of` doesn't seem to work on `NodeList` in iOS 9 Safari. I'll change it to regular `for` loop.

One test is failing only on iOS 11.3 Safari now. Looking into it.

Reply:
New version pushed. All tests should be passing now (at least passed when manually tested all platforms: Chrome, FF, IE11, Edge, Safari, iOS 9/11/12 Safari). Let's see what Travis says.